src/Core/Grid/Query/Monitoring/AbstractProductQueryBuilder.php

```diff
@@ -0,0 +1,188 @@
+     */
+    protected $contextShopId;
+
+    /**
+     * @var DoctrineSearchCriteriaApplicator
+     */
+    protected $searchCriteriaApplicator;
+
+    /**
+     * @var MultistoreContextCheckerInterface
+     */
+    protected $multistoreContextChecker;
+
+    /**
+     * @var FeatureInterface
+     */
+    protected $multistoreFeature;
+
+    /**
+     * @param Connection $connection
+     * @param string $dbPrefix
+     * @param DoctrineSearchCriteriaApplicator $searchCriteriaApplicator
+     * @param int $contextLangId
+     * @param int $contextShopId
+     * @param MultistoreContextCheckerInterface $multistoreContextChecker
+     * @param FeatureInterface $multistoreFeature
+     */
+    public function __construct(
+        Connection $connection,
+        $dbPrefix,
+        $contextLangId,
+        $contextShopId,
+        DoctrineSearchCriteriaApplicator $searchCriteriaApplicator,
+        MultistoreContextCheckerInterface $multistoreContextChecker,
+        FeatureInterface $multistoreFeature
+    ) {
+        parent::__construct($connection, $dbPrefix);
+        $this->contextLangId = $contextLangId;
+        $this->contextShopId = $contextShopId;
+        $this->searchCriteriaApplicator = $searchCriteriaApplicator;
+        $this->multistoreContextChecker = $multistoreContextChecker;
+        $this->multistoreFeature = $multistoreFeature;
+    }
+
+    /**
+     * Provides commonly reusable query for monitoring products lists
+     *
+     * @param SearchCriteriaInterface $searchCriteria
+     *
+     * @return QueryBuilder
+     */
+    protected function getProductsCommonQueryBuilder(SearchCriteriaInterface $searchCriteria)
+    {
+        $qb = $this->connection
+            ->createQueryBuilder()
+            ->select(['p.id_product', 'p.reference', 'p.active', 'pl.name'])
+            ->from($this->dbPrefix . 'product', 'p')
+            ->setParameter('context_lang_id', $this->contextLangId)
+            ->setParameter('context_shop_id', $this->contextShopId);
+
+        $qb->leftJoin(
+            'p',
+            $this->dbPrefix . 'product_lang',
+            'pl',
+            $this->multistoreFeature->isUsed() && $this->multistoreContextChecker->isSingleShopContext() ?
+                'p.id_product = pl.id_product AND pl.id_lang = :context_lang_id AND pl.id_shop = :context_shop_id' :
+                'p.id_product = pl.id_product AND pl.id_lang = :context_lang_id AND pl.id_shop = p.id_shop_default'
+        );
+
+        $qb->leftJoin(
+            'p',
+            $this->dbPrefix . 'product_shop',
+            'ps',
+            $this->multistoreFeature->isUsed() && $this->multistoreContextChecker->isSingleShopContext() ?
+                'p.id_product = ps.id_product AND ps.id_shop = :context_shop_id' :
+                'p.id_product = ps.id_product AND ps.id_shop = p.id_shop_default'
+        );
+
+        if ($this->multistoreFeature->isUsed() && $this->multistoreContextChecker->isSingleShopContext()) {
```

Comment:
You are using this condition multiple times, how about extracting it?
`$isSingleShopContext =  $this->multistoreFeature->isUsed() && $this->multistoreContextChecker->isSingleShopContext()`